File: app/src/main/java/pl/karol202/paintplus/tool/gradient/GradientProperties.java
package pl.karol202.paintplus.tool.gradient;

import android.os.Bundle;
import android.view.*;
import android.widget.AdapterView;
import android.widget.CheckBox;
import android.widget.CompoundButton;
import android.widget.Spinner;
import pl.karol202.paintplus.R;
import pl.karol202.paintplus.tool.ToolProperties;
import pl.karol202.paintplus.tool.gradient.ToolGradient.OnGradientEditListener;

public class GradientProperties extends ToolProperties implements OnGradientEditListener, AdapterView.OnItemSelectedListener, View.OnClickListener, GradientDialog.OnGradientUpdateListener, CompoundButton.OnCheckedChangeListener
{
	private ToolGradient toolGradient;
	private GradientShapes shapes;
	private GradientShapeAdapter adapterGradientShape;
	private GradientRepeatabilityAdapter adapterGradientRepeatability;
	
	private View view;
	private GradientPreviewView gradientPreview;
	private CheckBox checkGradientRevert;
	private Spinner spinnerGradientShape;
	private Spinner spinnerGradientRepeatability;
	
	@Override
	public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState)
	{
		super.onCreateView(inflater, container, savedInstanceState);
		setHasOptionsMenu(true);
		view = inflater.inflate(R.layout.properties_gradient, container, false);
		toolGradient = (ToolGradient) tool;
		toolGradient.setOnGradientEditListener(this);
		shapes = toolGradient.getShapes();
		adapterGradientShape = new GradientShapeAdapter(getActivity(), shapes.getShapes());
		adapterGradientRepeatability = new GradientRepeatabilityAdapter(getActivity());
		
		gradientPreview = view.findViewById(R.id.gradient_preview);
		gradientPreview.setGradient(toolGradient.getGradient());
		gradientPreview.setOnClickListener(this);
		
		checkGradientRevert = view.findViewById(R.id.check_gradient_revert);
		checkGradientRevert.setChecked(toolGradient.isReverted());
		checkGradientRevert.setOnCheckedChangeListener(this);
		
		spinnerGradientShape = view.findViewById(R.id.spinner_gradient_shape);
		spinnerGradientShape.setAdapter(adapterGradientShape);
		spinnerGradientShape.setSelection(shapes.getIdOfShape(toolGradient.getShape()));
		spinnerGradientShape.setOnItemSelectedListener(this);
		
		spinnerGradientRepeatability = view.findViewById(R.id.spinner_gradient_repeatability);
		spinnerGradientRepeatability.setAdapter(adapterGradientRepeatability);
		spinnerGradientRepeatability.setSelection(toolGradient.getRepeatability().ordinal());
		spinnerGradientRepeatability.setOnItemSelectedListener(this);
		
		return view;
	}
	
	@Override
	public void onCreateOptionsMenu(Menu menu, MenuInflater inflater)
	{
		if(toolGradient.isInEditMode()) inflater.inflate(R.menu.menu_tool_gradient, menu);
	}
	
	@Override
	public boolean onOptionsItemSelected(MenuItem item)
	{
		int id = item.getItemId();
		if(toolGradient.isInEditMode())
		{
			switch(id)
			{
			case R.id.action_apply:
				toolGradient.apply();
				break;
			case R.id.action_cancel:
				toolGradient.cancel();
				break;
			}
			getActivity().invalidateOptionsMenu();
			return true;
		}
		return super.onOptionsItemSelected(item);
	}
	
	@Override
	public void onGradientSet()
	{
		getActivity().invalidateOptionsMenu();
	}
	
	@Override
	public void onClick(View v)
	{
		GradientDialog dialog = new GradientDialog(getActivity(), toolGradient.getGradient());
		dialog.setGradientUpdateListener(this);
		dialog.show();
	}
	
	@Override
	public void onGradientUpdated()
	{
		gradientPreview.update();
	}
	
	@Override
	public void onItemSelected(AdapterView<?> parent, View view, int position, long id)
	{
		if(parent == spinnerGradientShape)
			toolGradient.setShape(shapes.getShape(position));
		else if(parent == spinnerGradientRepeatability)
			toolGradient.setRepeatability(GradientRepeatability.values()[position]);
	}
	
	@Override
	public void onNothingSelected(AdapterView<?> parent)
	{
		System.err.println("nothing!");
	}
	
	@Override
	public void onCheckedChanged(CompoundButton compoundButton, boolean b)
	{
		toolGradient.setRevert(b);
	}
}

File: app/src/main/java/pl/karol202/paintplus/tool/gradient/GradientProperties.kt
/*
 * Copyright 2017 karol-202
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */
package pl.karol202.paintplus.tool.gradient

import android.os.Bundle
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import androidx.fragment.app.Fragment
import androidx.lifecycle.lifecycleScope
import org.koin.android.ext.android.inject
import org.koin.androidx.viewmodel.ext.android.sharedViewModel
import pl.karol202.paintplus.R
import pl.karol202.paintplus.color.picker.ColorPickerConfig
import pl.karol202.paintplus.databinding.PropertiesGradientBinding
import pl.karol202.paintplus.tool.gradient.shape.GradientShapeAdapter
import pl.karol202.paintplus.util.collectIn
import pl.karol202.paintplus.util.setOnItemSelectedListener
import pl.karol202.paintplus.util.viewBinding
import pl.karol202.paintplus.viewmodel.PaintViewModel

class GradientProperties : Fragment(R.layout.properties_gradient)
{
	private val paintViewModel by sharedViewModel<PaintViewModel>()
	private val toolGradient by inject<ToolGradient>()

	private val views by viewBinding(PropertiesGradientBinding::bind)

	override fun onViewCreated(view: View, savedInstanceState: Bundle?)
	{
		setHasOptionsMenu(true)

		toolGradient.updateEventFlow.collectIn(lifecycleScope) { activity?.invalidateOptionsMenu() }

		views.gradientPreview.setGradient(toolGradient.gradient)
		views.gradientPreview.setOnClickListener { openGradientDialog() }

		views.checkGradientRevert.isChecked = toolGradient.isReverted
		views.checkGradientRevert.setOnCheckedChangeListener { _, checked -> toolGradient.isReverted = checked }

		views.spinnerGradientShape.adapter = GradientShapeAdapter(requireContext(), toolGradient.gradientShapes)
		views.spinnerGradientShape.setSelection(toolGradient.gradientShapes.indexOf(toolGradient.shape))
		views.spinnerGradientShape.setOnItemSelectedListener { toolGradient.shape = toolGradient.gradientShapes[it] }

		views.spinnerGradientRepeatability.adapter = GradientRepeatabilityAdapter(requireContext())
		views.spinnerGradientRepeatability.setSelection(toolGradient.repeatability.ordinal)
		views.spinnerGradientRepeatability.setOnItemSelectedListener {
			toolGradient.repeatability = GradientRepeatability.values()[it]
		}
	}

	override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater)
	{
		if(toolGradient.isInEditMode) inflater.inflate(R.menu.menu_tool_gradient, menu)
	}

	override fun onOptionsItemSelected(item: MenuItem): Boolean
	{
		if(toolGradient.isInEditMode)
		{
			when(item.itemId)
			{
				R.id.action_apply -> toolGradient.apply()
				R.id.action_cancel -> toolGradient.cancel()
			}
			activity?.invalidateOptionsMenu()
			return true
		}
		return super.onOptionsItemSelected(item)
	}

	private fun openGradientDialog() =
			GradientDialog(requireContext(), toolGradient.gradient, this::pickColor, this::onGradientEdited).show()

	private fun pickColor(config: ColorPickerConfig, callback: (Int?) -> Unit) =
			paintViewModel.makeActionRequest(PaintViewModel.ActionRequest.PickColor(config, callback))

	private fun onGradientEdited(gradient: Gradient)
	{
		toolGradient.gradient = gradient
		views.gradientPreview.setGradient(gradient)
	}
}
